Update Avi-On to work with latest API

antsar opened the pull request:
## Description:

The `light.avion` component supports two ways of configuration: username/password (a web API is used to retrieve all device metadata), or manually enumerating devices by MAC address and ID. The Avi-On web API has changed, and the Home Assistant integration can no longer fetch devices. This PR updates Home Assistant to work with the new API.

**Related issue (if applicable):** fixes #17779

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7073

## Example entry for `configuration.yaml` (if applicable):
This PR does not impact configuration. See current example config [here](https://www.home-assistant.io/components/light.avion).

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - ~Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)~ **N/A**

If the code communicates with devices, web services, or third-party tools:
  - ~New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).~ **N/A**
  - ~New dependencies are only imported inside functions that use them ([example][ex-import]).~ **N/A**
  - ~New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.~ **N/A**
  - ~New files were added to `.coveragerc`.~ **N/A**

If the code does not interact with devices:
  - ~Tests have been added to verify that the new code works.~ **N/A**

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

antsar committed:
Update Avi-On to work with the API.

fabaff commented on homeassistant/components/light/avion.py:
```diff
@@ -10,14 +10,14 @@
 import homeassistant.helpers.config_validation as cv
 
-REQUIREMENTS = ['avion==0.7']
+REQUIREMENTS = ['antsar-avion==0.9']
```
Why are you changing the requirement?

antsar replied:
@fabaff thanks for noticing/asking. I'm curious if I've gone about this the right way.

I've opened a PR making the necessary changes to `avion`. Since this vendor API change breaks the library, I thought I'd switch to my fork in the meantime to keep Home Assistant working. The author seems like a busy guy (my last PR took a few months to get merged).

`avion` PR: https://github.com/mjg59/python-avion/pull/9
Previous `avion` PR : https://github.com/mjg59/python-avion/pull/8 (merged, but I later realized the fix was incomplete)

Is switching a bad way to handle this? If it's preferable to wait for the `avion` PR to get merged, I'm ok with that.

fabaff replied:
> Is switching a bad way to handle this?

No, can be done especially If bugs are not addressed or the developer/maintainer is non-responsive. If @mjg59 is busy then we can go on with your fork. Would be nice if you could address those two things in your fork:

- https://github.com/mjg59/python-avion/issues/10
- https://github.com/mjg59/python-avion/issues/11

Also, I would like to suggest that you update the README in your fork with to make it clear that's a fork.

MartinHjelmare replied:
@antsar Instead of changing to your fork you could ask to be added as a collaborator to the main repo. Of course this depends on how @mjg59 feels about this.

antsar committed:
Use voluptuous defaults instead of .get()

antsar committed:
Bump library version.

fabaff commented on homeassistant/components/light/avion.py:
```diff
@@ -26,12 +26,13 @@
 DEVICE_SCHEMA = vol.Schema({
     vol.Optional(CONF_NAME): cv.string,
```
I think that we should provide a default name
```python
DEFAULT_NAME = 'Avi-On'
...
    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
```

antsar replied:
The library takes care of this already, so providing our own default name seems redundant.

https://github.com/antsar/python-avion/blob/0.9.1/avion/__init__.py#L88

fabaff commented on homeassistant/components/light/avion.py:
```diff
@@ -42,24 +43,17 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-                                     mac[2], mac[3])
-                lights.append(AvionLight(device))
+        devices = avion.get_devices(
+            config[CONF_USERNAME], config[CONF_PASSWORD])
```
Use `.get()`. No need to define a default in the configuration validation for the username and the password as `None` is the [default](https://docs.python.org/3.8/library/stdtypes.html#dict.get) already. 

Thus I don't agree with [this comment](https://github.com/home-assistant/home-assistant/pull/17780#pullrequestreview-168204198). Sometimes you want the `KeyError` to happen.

antsar replied:
This is inside a conditional block which ensures those keys exist, so I don't think `.get()` is necessary.

```python
     if CONF_USERNAME in config and CONF_PASSWORD in config:
         devices = avion.get_devices(
             config[CONF_USERNAME], config[CONF_PASSWORD])
```

antsar committed:
Remove unnecessary voluptuous defaults. Fix manually-configured devices.

API-discovered devices are already Avion objects, but manually-configured devices need to be instantiated as Avion objects first.

antsar commented on homeassistant/components/light/avion.py:
```diff
@@ -42,24 +43,17 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
     for address, device_config in config[CONF_DEVICES].items():
         device = {}
         device['name'] = device_config[CONF_NAME]
```
[Fixed](https://github.com/home-assistant/home-assistant/pull/17780/commits/d52f61b16b795489f02d4ed45410ecdc0cd623c3#diff-5546f05751d025749fd1b4969a0ab0beR55). Thank you for catching this.

antsar committed:
Use .get() where appropriate.

fabaff committed:
Remove default
